### DISCUSSION TITLE
One python LSP running but multiple check statements

### DISCUSSION THREAD

**Discussion by speri203** [LazyVim/LazyVim#447] [General]
I have been using pyright for a while and wanted to try out the ruff-lsp for python development. I have gone ahead and uninstalled pyright via mason and have installed ruff-lsp but I seem to be getting multiple type checker statements for a given row instead of just one. For example, in the following screenshot, I see that the os import is not used, but there are two debugging statements saying the same thing. If I remove ruff-lsp all together then only the one in green (the first one) is displayed. Is there an LSP running in the background I am not seeing? I have also tried uninstalling the flake8 linter with nothing changing.
![Screenshot 2023-03-17 at 1 25 18 PM](https://user-images.githubusercontent.com/36865599/225846708-1275d374-9d6f-4664-a462-61ffeede2f52.png)

The following LSPs are running on this particular file
![Screenshot 2023-03-17 at 1 25 38 PM](https://user-images.githubusercontent.com/36865599/225847065-f0409966-594f-4078-860b-71580769e622.png)

For reference, this is what my lsp.lua looks like this:
```
return {
  {
    "williamboman/mason.nvim",
    opts = function(_, opts)
      vim.list_extend(opts.ensure_installed, {
        --"pyright",
        "ruff-lsp",
        --"black",
        "stylua",
        "shfmt",
        "isort",
        "prettier",
        "gomodifytags",
        "gofumpt",
        "iferr",
        "impl",
        "goimports",
        "gopls",
        "blue",
      })
    end,
  },

  -- Setup null-ls with `black`
  {
    "jose-elias-alvarez/null-ls.nvim",
    opts = function(_, opts)
      local nls = require("null-ls")
      opts.sources = vim.list_extend(opts.sources, {
        --nls.builtins.formatting.black,
        nls.builtins.formatting.blue,
        nls.builtins.formatting.stylua,
        nls.builtins.formatting.prettier,
        nls.builtins.formatting.gofmt,
        nls.builtins.formatting.isort,
      })
    end,
  },
  {
    "neovim/nvim-lspconfig",
    dependencies = {
      "b0o/SchemaStore.nvim",
      version = false, -- last release is way too old
    },
    opts = {
      -- make sure mason installs the server
      servers = {
        jsonls = {
          -- lazy-load schemastore when needed
          on_new_config = function(new_config)
            new_config.settings.json.schemas = new_config.settings.json.schemas or {}
            vim.list_extend(new_config.settings.json.schemas, require("schemastore").json.schemas())
          end,
          settings = {
            json = {
              format = {
                enable = true,
              },
              validate = { enable = true },
            },
          },
        },
      },
    },
  },
  {
    "jay-babu/mason-nvim-dap.nvim",
    opts = {
      ensure_installed = { "python", "delve" },
    },
  },
  {
    "linux-cultist/venv-selector.nvim",
    opts = {},
    keys = { { "<leader>lv", "<cmd>:VenvSelect<cr>", desc = "Select VirtualEnv" } },
  },
}
```

**Comment by folke** [comment]
did you remove flake8 from null-ls? Not sure if that's causing it though. I don't use python

**Reply by speri203** [reply]
Yeah, I did but I am not sure where the other linting comments are coming from, to be honest. Maybe pyright? I'll have to do experiments.

**Reply by dhruvinsh** [reply]
looking at the code flake8 is not remove yet

``` lua
    "jose-elias-alvarez/null-ls.nvim",
    opts = function(_, opts)
      local nls = require("null-ls")
      opts.sources = vim.list_extend(opts.sources, {
        --nls.builtins.formatting.black,
        nls.builtins.formatting.blue,
        nls.builtins.formatting.stylua,
        nls.builtins.formatting.prettier,
        nls.builtins.formatting.gofmt,
        nls.builtins.formatting.isort,
      })
    end,
```

Here basically you are extending opts.sources. But flake8 is part of standard layvim system. you need to remove flake8 from opt.sources.

Add something like this,
``` lua
opts.sources = vim.tbl_filter(function(source)
   return source.name ~= "flake8"
end, opts.sources)
```

@folke if it is not too much to ask, and I also read that you do not use pytho language, can you please remove `flake8` from the package system and provide as an extra like you do for javascript. Lot of people in python community started migrate to `ruff`

**Reply by amaanq** [reply]
+1 for ruff/ruff-lsp

**Reply by folke** [reply]
Just pushed an update to remove `flake8`

**Reply by fredrikaverpil** [reply]
@folke forgive me, but I'm both new to neovim and lazyvim... what do I do to get this update you mentioned pushing?

**Reply by speri203** [reply]
> @folke forgive me, but I'm both new to neovim and lazyvim... what do I do to get this update you mentioned pushing?

"LazyVim" is managed by lazy.vim and you can see it as a plugin if you do leader+l (as in lake) followed by "/" for searching and typing in LazyVim. To update it you can open up this same dialogue and press Shift+u and it will fetch any updates if there are any. This is also how you would update all your other plugins.
